Feature/survey upload interface

OBaddeley opened the pull request:
[Story](https://trello.com/c/CInaJ9PP/71-interface-for-uploading-and-looking-at-surveys)

This PR:
- provides an interface to upload a survey file.
- uses a background job to process the file

Something to be aware of, with the job running with sidekiq and the files being saved to the app/tmp file, Sidekiq and the App will need to be running on the same instance. 

![Screen Shot 2020-06-11 at 12 45 23 PM](https://user-images.githubusercontent.com/10471021/84332602-70ff6b00-abe1-11ea-8e5b-c7831a7c53d4.png)

mermop commented on .rubocop-https---bitbucket-org-rabidtech-rabid-dotfiles-raw-master--rubocop-yml:
```diff
@@ -162,55 +162,3 @@ Style/FormatStringToken:
 Lint/AmbiguousBlockAssociation:
   Exclude:
     - "spec/**/*"
-
```
Can we `.gitignore` this file? Keeps changing and clogging up unrelated PRs

mermop commented on app/services/survey_import_service.rb:
```diff
@@ -13,37 +15,62 @@ def import_item_name
   end
 
+  def attachable
+    @file.rewind
```
Can you explain why this is necessary?

OBaddeley replied:
Because it is an IO string that gets open (and read in the job) so it needs to be rewound

mermop commented on config/routes.rb:
```diff
@@ -66,9 +67,9 @@
   # If you want the Sidekiq web console in production environments you need to
   # put it behind some authentication first.
   #
+  require "sidekiq/web"
```
Why have you moved this out of the development section? We'll need to change the authentication - currently the username and password for the sidekiq console is committed to the repo

OBaddeley replied:
Because it did not work in development inside that block

OBaddeley replied:
because it has to be require to check if defined I think

mermop replied:
Ahhhh I see - so it's not being mounted outside of development, but it's still being required.

One option would be to just remove the `if defined?(Sidekiq::Web)` and have 

```
if Rails.env.development?
  require "sidekiq/web"
  mount Sidekiq::Web => "/sidekiq" # Sidekiq monitoring console
end
```

Not 100% that works

mermop commented on app/services/survey_import_service.rb:
```diff
@@ -13,37 +15,62 @@ def import_item_name
+
+  def content_type
+    # Marcel is the content type detection library used by Active Storage
+    Marcel::MimeType.for @file, name: filename
```
```suggestion
    Marcel::MimeType.for(@file, name: filename)
```

Just for consistency

mermop commented:
Is there any way that we can avoid the need for Sidekiq and the app running on the same instance? Just thinking about Docker..

OBaddeley commented:
Potentially if we make a SurveyUpload Model and then save the file (and maybe the consultation) against that, and then pass the model instance through to the Job and then the importer service etc, then might work ?

mermop commented:
Sounds good to me, then we can just save and access the files using the same ActiveStorage pattern as other file saving in the application

OBaddeley commented:
- added the `SurveyImport` model and made adjustments for that. 
- updated upload view as per Meron's recommendations
- populate filename based on submission file or survey file
- update filename filter to work with surveys 

<img width="1128" alt="Screen Shot 2020-06-15 at 5 11 14 PM" src="https://user-images.githubusercontent.com/10471021/84620861-34a77400-af2d-11ea-942d-1c6fc8c518b9.png">

<img width="682" alt="Screen Shot 2020-06-15 at 11 51 41 AM" src="https://user-images.githubusercontent.com/10471021/84620886-45f08080-af2d-11ea-9ca1-b56ca3f85205.png">

<img width="503" alt="Screen Shot 2020-06-15 at 5 26 22 PM" src="https://user-images.githubusercontent.com/10471021/84620930-64567c00-af2d-11ea-8560-459699e3d8c6.png">

<img width="418" alt="Screen Shot 2020-06-15 at 5 26 32 PM" src="https://user-images.githubusercontent.com/10471021/84620943-6e787a80-af2d-11ea-8133-113070f14c31.png">

mermop commented on .gitignore:
```diff
@@ -40,3 +40,6 @@ yarn-debug.log*
 /.env
+
+# ignore rubocop diffs
+/.rubocop-https---bitbucket-org-rabidtech-rabid-dotfiles-raw-master--rubocop-yml
```
I think you might also need to delete the file, the diff is still showing up

mermop commented on app/controllers/surveys_controller.rb:
```diff
@@ -0,0 +1,32 @@
+
+    SurveyImportJob.perform_later(@survey_import)
+
+    notice = "Survey uploaded successfully"
```
What happens if you upload the wrong type of file - does it fail quietly?

mermop replied:
Also - someone seeing this notice might be confused that they're not actually seeing the submissions yet. Could we have something like "Survey uploaded, awaiting processing" (or a friendlier version of that)?

OBaddeley replied:
The accept file upload types prevents other file types being uploaded from the view
`<%= form.file_field :file, accept: survey_upload_file_types, hide_label: true %>`

Updated copy.

mermop commented on app/models/submission.rb:
```diff
@@ -30,7 +30,14 @@ class Submission < ApplicationRecord
   scope :active, -> { where.not(state: "archived") }
   scope :search_by_filename, lambda { |query|
-    joins(file_attachment: :blob).where("active_storage_blobs.filename ILIKE ?", "%#{query}%")
+    sql1 = joins(survey: [original_file_attachment: :blob])
```
Can you give these subqueries a more descriptive name? We could even make this two scopes to avoid it being so large

mermop commented on app/jobs/survey_import_job.rb:
```diff
@@ -0,0 +1,7 @@
+class SurveyImportJob < ApplicationJob
+  queue_as :default
```
Does it need this line?

OBaddeley replied:
Think it is nice to be explicit, but maybe not ?

mermop commented on app/models/submission.rb:
```diff
@@ -89,10 +96,14 @@ def add_tag(params)
   end
 
   def filename
+    return survey.original_file.filename if survey && survey&.original_file&.attached?
```
It feels like there's a lot going on in this line...
Could we give the `survey` a `title` or `filename` and delegate some of this logic there? Like -

```
# models/submission.rb
def filename
  return survey.filename if survey && survey.filename
end

# models/survey.rb
def filename
  return original_file.filename if original_file&.attached?
  # ??? I guess nil otherwise? Why wouldn't a survey have an original file attached?
end
```

mermop commented:
The text doesn't look right here, it doesn't have any question tokens - is that a new problem? 

<img width="149" alt="Screen Shot 2020-06-16 at 10 10 19 AM" src="https://user-images.githubusercontent.com/1615322/84711078-9f05f600-afb9-11ea-9f2d-d258faf994a5.png">

OBaddeley commented:
Oh, no it is because I took a screen shot of an old file... before survey questions where doing there thing properly.

![Screen Shot 2020-06-16 at 10 33 05 AM](https://user-images.githubusercontent.com/10471021/84712496-daee8a80-afbc-11ea-8219-80ebcd98501d.png)

mermop commented on app/models/submission.rb:
```diff
@@ -30,22 +30,15 @@ class Submission < ApplicationRecord # rubocop:disable Metrics/ClassLength
+        "active_storage_blobs.filename ILIKE :query OR "\
+        "blobs_active_storage_attachments.filename ILIKE :query",
+        query: "%#{query}%"
+      )
```
🙌

OBaddeley replied:
option 3 👍  :)

mermop replied:
Hooray!
